# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Clear11.cpp
# Commit: 754436d8b3a2
# Full Hash: 754436d8b3a2023b90eb247b3a70335e74f2fdf6
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-05-09 09:14:23
# Regressor Bug: 1633628
# File Overlap Count: 3
# Description:
#   Bug 1633628 - Vender: Don't use ClearView if we previously used dual source blending on Intel gen6. r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D74137
# ==============================================================================

diff -r f8fd6f0a5102 -r 754436d8b3a2 gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Clear11.cpp
--- a/gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Clear11.cpp	Fri May 08 18:39:28 2020 +0000
+++ b/gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Clear11.cpp	Fri May 08 19:28:53 2020 +0000
@@ -505,7 +505,36 @@
         const auto &framebufferRTV = renderTarget->getRenderTargetView();
         ASSERT(framebufferRTV.valid());
 
-        if ((!(mRenderer->getRenderer11DeviceCaps().supportsClearView) && needScissoredClear) ||
+        bool canClearView = true;
+        if (mRenderer->getFeatures().emulateClearViewAfterDualSourceBlending.enabled) {
+            // Check the current state to see if we were using dual source blending
+            auto isDualSource = [](auto blend) { switch (blend) {
+                        case D3D11_BLEND_SRC1_COLOR:
+                        case D3D11_BLEND_INV_SRC1_COLOR:
+                        case D3D11_BLEND_SRC1_ALPHA:
+                        case D3D11_BLEND_INV_SRC1_ALPHA:
+                                return true;
+                        default:
+                                return false;
+            }};
+            FLOAT blendFactor[4];
+            UINT sampleMask;
+            ID3D11BlendState *blendState;
+            D3D11_BLEND_DESC blendDesc;
+            deviceContext1->OMGetBlendState(&blendState, blendFactor, &sampleMask);
+            if (blendState) {
+                    blendState->GetDesc(&blendDesc);
+                    // You can only use dual source blending on slot 0 so only check there
+                    if (isDualSource(blendDesc.RenderTarget[0].SrcBlend) ||
+                        isDualSource(blendDesc.RenderTarget[0].DestBlend) ||
+                        isDualSource(blendDesc.RenderTarget[0].SrcBlendAlpha) ||
+                        isDualSource(blendDesc.RenderTarget[0].DestBlendAlpha)) {
+                            canClearView = false;
+                    }
+            }
+        }
+
+        if ((!(mRenderer->getRenderer11DeviceCaps().supportsClearView && canClearView) && needScissoredClear) ||
             clearParams.colorType != GL_FLOAT ||
             (formatInfo.redBits > 0 && !clearParams.colorMaskRed) ||
             (formatInfo.greenBits > 0 && !clearParams.colorMaskGreen) ||